pkg/reconciler/revision/background.go

```diff
@@ -0,0 +1,260 @@
+// workItem for each container we need to resolve for the overall result.
+type resolveResult struct {
+	statuses  []v1.ContainerStatus
+	err       error
+	remaining int
+
+	opt              k8schain.Options
+	registriesToSkip sets.String
+
+	completionCallback func()
+}
+
+// workItem is a single task submitted to the queue, to resolve a single image
+// for a resolveResult.
+type workItem struct {
+	result  *resolveResult
+	timeout time.Duration
+
+	name  string
+	image string
+	index int
+}
+
+func newBackgroundResolver(logger *zap.SugaredLogger, resolver imageResolver, enqueue func(types.NamespacedName)) *backgroundResolver {
+	r := &backgroundResolver{
+		logger: logger,
+
+		resolver: resolver,
+		enqueue:  enqueue,
+
+		results: make(map[types.NamespacedName]*resolveResult),
+		queue:   workqueue.NewNamed("digests"),
+	}
+
+	return r
+}
+
+// Start starts the worker threads and runs maxInFlight workers until the stop
+// channel is closed. It returns a done channel which will be closed when all
+// workers have exited.
+func (r *backgroundResolver) Start(stop <-chan struct{}, maxInFlight int) (done chan struct{}) {
+	var wg sync.WaitGroup
+
+	// Run the worker threads.
+	wg.Add(maxInFlight)
+	for i := 0; i < maxInFlight; i++ {
+		go func() {
+			defer wg.Done()
+			for {
+				item, shutdown := r.queue.Get()
+				if shutdown {
+					return
+				}
+
+				rrItem, ok := item.(*workItem)
+				if !ok {
+					r.logger.Fatalf("Unexpected work item type: want: %s, got: %s\n",
+						reflect.TypeOf(&workItem{}).Name(), reflect.TypeOf(item).Name())
+				}
+
+				r.processWorkItem(rrItem)
+			}
+		}()
+	}
+
+	// Shut down the queue once the stop channel is closed, this will cause all
+	// the workers to exit.
+	go func() {
+		<-stop
+		r.queue.ShutDown()
+	}()
+
+	// Return a done channel which is closed once all workers exit.
+	done = make(chan struct{})
+	go func() {
+		wg.Wait()
+		close(done)
+	}()
+
+	return done
+}
+
+// Resolve is intended to be called from a reconciler to resolve a revision. If
+// the resolver already has the digest in cache it is returned immediately, if
+// it does not and no resolution is already in flight a resolution is triggered
+// in the background.
+// If this method returns nil, nil this implies a resolve was triggered or is
+// already in progress, so the reconciler should exit and wait for the revision
+// to be re-enqueued when the result is ready.
+func (r *backgroundResolver) Resolve(rev *v1.Revision, opt k8schain.Options, registriesToSkip sets.String, timeout time.Duration) ([]v1.ContainerStatus, error) {
+	r.mu.Lock()
+	defer r.mu.Unlock()
+
+	name := types.NamespacedName{
+		Name:      rev.Name,
+		Namespace: rev.Namespace,
+	}
+
+	result, inFlight := r.results[name]
+	if !inFlight {
+		r.addWorkItems(rev, opt, registriesToSkip, timeout)
+		return nil, nil
+	}
+
+	if !result.ready() {
+		return nil, nil
+	}
+
+	return r.results[name].statuses, r.results[name].err
```

Comment:
Should we be clearing here and skip resolution if the revision has the relevant status set already to avoid this becoming a memory hog?

Reply:
Im clearing in the reconcile at the moment, so Resolve() never actually gets called if we've already got ContainerStatuses on the revision. Could certainly move the Clear() in to the Resolve, though (although then Resolve becomes a little less idempotent, and we still need Clear for the delete case I think).

Reply:
Oh, I haven't looked as far sorry. I guess that works too, yes!